cpp/include/raft/cluster/kmeans_balanced.cuh

```diff
@@ -0,0 +1,341 @@
+  raft::cluster::kmeans_balanced::fit(handle, params, X, centroids, mapping_op);
+  raft::cluster::kmeans_balanced::predict(handle, params, X, centroids_const, labels, mapping_op);
+}
+
+/**
+ * @brief Randomly initialize centers and apply expectation-maximization-balancing iterations
+ *
+ * This is essentially the non-hierarchical balanced k-means algorithm which is used by the
+ * hierarchical algorithm once to build the mesoclusters and once per mesocluster to build the fine
+ * clusters.
+ *
+ * @code{.cpp}
+ *   #include <raft/core/handle.hpp>
+ *   #include <raft/cluster/kmeans_balanced.cuh>
+ *   #include <raft/cluster/kmeans_balanced_types.hpp>
+ *   ...
+ *   raft::handle_t handle;
+ *   raft::cluster::kmeans_balanced_params params;
+ *   auto centroids = raft::make_device_matrix<float, int>(handle, n_clusters, n_features);
+ *   auto labels = raft::make_device_vector<int, int>(handle, n_samples);
+ *   auto sizes = raft::make_device_vector<int, int>(handle, n_clusters);
+ *   raft::cluster::kmeans_balanced::build_clusters(handle, params, X, centroids, labels, sizes);
+ * @endcode
+ *
+ * @tparam DataT Type of the input data.
+ * @tparam MathT Type of the centroids and mapped data.
+ * @tparam IndexT Type used for indexing.
+ * @tparam LabelT Type of the output labels.
+ * @tparam CounterT Counter type supported by CUDA's native atomicAdd.
+ * @tparam MappingOpT Type of the mapping function.
+ * @param[in]  handle        The raft handle
+ * @param[in]  params        Structure containing the hyper-parameters
+ * @param[in]  X             Training instances to cluster. The data must be in row-major format.
+ *                           [dim = n_samples x n_features]
+ * @param[out] centroids     The output centroids [dim = n_clusters x n_features]
+ * @param[out] labels        The output labels [dim = n_samples]
+ * @param[out] cluster_sizes Size of each cluster [dim = n_clusters]
+ * @param[in]  mapping_op    (optional) Functor to convert from the input datatype to the
+ *                           arithmetic datatype. If DataT == MathT, this must be the identity.
+ * @param[in]  mr            (optional) Device memory resource.
+ * @param[in]  X_norm        (optional) Dataset's row norms [dim = n_samples]
+ */
+template <typename DataT,
+          typename MathT,
+          typename IndexT,
+          typename LabelT,
+          typename CounterT,
+          typename MappingOpT>
+void build_clusters(handle_t const& handle,
+                    const kmeans_balanced_params& params,
+                    raft::device_matrix_view<const DataT, IndexT> X,
+                    raft::device_matrix_view<MathT, IndexT> centroids,
+                    raft::device_vector_view<LabelT, IndexT> labels,
+                    raft::device_vector_view<CounterT, IndexT> cluster_sizes,
+                    MappingOpT mapping_op               = raft::identity_op(),
+                    rmm::mr::device_memory_resource* mr = nullptr,
```

Comment:
@cjnolet This function is called by IVF-PQ which creates its own temporary pool resources (managed and device) with a controlled initial size. We can't really / don't want to change the global resource, right? Nor can we rely on it being set as we want.

We should probably not be exposing `build_clusters` nor `calc_centers_and_sizes` in the public API at all but the problem is that they are used by `knn` so it would break encapsulation to have them only in `cluster::detail`.

I'm not sure what to do. Either we expose the resource, or we allow `knn` to compromise on encapsulation. Unless you know a better way to pass the temporary pool from IVF-PQ to this function?

Reply:
We run into issues when we're creating our own pools "behind the scenes" which aren't directly controllable by the user. We can always provide documentation to help the power user make more informed decisions but when users are building a workflow and computing the amount of memory needed to make their calls into RAFT (or the downstream libraries like cuml), they most often will 
1) just use the default RMM resource, 
2) set a global pool to get a speedup (which most often will be some large percentage of their available GPU memory), or 
3) set a managed / managed pool resource to get the benefits of the pool but oversubscribe their GPU to avoid the OOMs. 

If users are invoking several RAFT functions in a row (such as many cuml algorithms do), it's going to make memory management very complicated, if not close to impossible, when we're creating resources and allocating pools behind their back. Thus, we need to afford them the flexibility of being able to specify these things so, for example, the user who sets a pool w/ 90% of their available memory isn't scratching their head when they go to compute the nearest neighbors and encounter an OOM. It's much more straightforward to suggest they can get a perf boost by setting a pooled allocator for the `workspace_resource()` than to tell them they have no control over this.

Reply:
Oh wait, you're suggesting that IVF-PQ shouldn't set its own pool?

Reply:
> Oh wait, you're suggesting that IVF-PQ shouldn't set its own pool?

Yes, I am suggesting that for the longer-term. 

In the near-term, I propose that we add a copy constructor to the handle which accepts an optional `workspace_resource` and does a shallow-copy of the underlying array of existing resources, setting the new resource on the new instance. This will be useful to users so they can invoke different functions that can make use of different pool resources but still give them the control over their configuration (essentially I'm just saying we remove the explicit `mr` argument from this function and use the one already on the `handle_t`, whilst having the existing IVF-pq implementation use a new `handle_t` if it's going to use a new pool). Eventually we will also enable the user to control the resources in ivf-pq as well.

Reply:
In the longer longer term, I'd like for _all_ of RAFT's primitive functions to use the workspace_resource on the given `handle_t`, if one has been set (and fall back to the default resource set on the device). But that's probably a ways out.